File: src/scaffolds/backends/fastapi/services/log_service.py
"""操作日志服务"""
from utils.database import Database
from utils.response import ResultCode
from services.user_service import BizError, _format_time


class LogService:

    @staticmethod
    def page_query(query: dict):
        """分页查询操作日志（带条件）"""
        page_num = query.get('pageNum', 1)
        page_size = query.get('pageSize', 10)
        offset = (page_num - 1) * page_size

        where, params = 'WHERE 1=1', []
        if query.get('username'):
            where += ' AND username LIKE %s'; params.append(f"%{query['username']}%")
        if query.get('operation'):
            where += ' AND operation LIKE %s'; params.append(f"%{query['operation']}%")
        if query.get('startTime'):
            where += ' AND create_time >= %s'; params.append(query['startTime'])
        if query.get('endTime'):
            where += ' AND create_time <= %s'; params.append(query['endTime'])

        order_col = {'username': 'username', 'operation': 'operation', 'executeTime': 'execute_time',
                     'createTime': 'create_time'}.get(query.get('orderBy'), 'create_time')
        direction = 'ASC' if query.get('order') == 'asc' else 'DESC'

        total = Database.execute_one(f'SELECT COUNT(*) as total FROM operation_log {where}', tuple(params))['total']
        records = Database.execute_query(
            f'SELECT * FROM operation_log {where} ORDER BY {order_col} {direction} LIMIT %s OFFSET %s',
            tuple(params + [page_size, offset]))
        for r in records:
            _format_time(r)

        return {'records': records, 'total': total, 'pageNum': page_num, 'pageSize': page_size}

    @staticmethod
    def list_all():
        """查询所有操作日志"""
        records = Database.execute_query('SELECT * FROM operation_log ORDER BY create_time DESC')
        for r in records:
            _format_time(r)
        return records

    @staticmethod
    def get_by_id(log_id: int):
        """根据ID查询操作日志"""
        log = Database.execute_one('SELECT * FROM operation_log WHERE id = %s', (log_id,))
        if not log:
            raise BizError(ResultCode.NOT_FOUND, '日志不存在')
        _format_time(log)
        return log

    @staticmethod
    def delete_by_id(log_id: int):
        """删除操作日志"""
        Database.execute_update('DELETE FROM operation_log WHERE id = %s', (log_id,))

    @staticmethod
    def delete_batch(ids: list):
        """批量删除操作日志"""
        if not ids:
            raise BizError(ResultCode.BAD_REQUEST, '参数错误')
        placeholders = ','.join(['%s'] * len(ids))
        Database.execute_update(f'DELETE FROM operation_log WHERE id IN ({placeholders})', tuple(ids))

File: src/scaffolds/backends/go/services/log_service.go
// Package services 操作日志服务
package services

import (
	"fmt"
	"go-mysql-backend/config"
	"go-mysql-backend/models"
	"go-mysql-backend/utils"
	"strings"
)

// LogPageQuery 分页查询操作日志
func LogPageQuery(query models.PageQuery) (map[string]interface{}, error) {
	query.Defaults()
	offset := (query.PageNum - 1) * query.PageSize

	where, args := "WHERE 1=1", []interface{}{}
	if query.Username != "" {
		where += " AND username LIKE ?"; args = append(args, "%"+query.Username+"%")
	}
	if query.Operation != "" {
		where += " AND operation LIKE ?"; args = append(args, "%"+query.Operation+"%")
	}
	if query.StartTime != "" {
		where += " AND create_time >= ?"; args = append(args, query.StartTime)
	}
	if query.EndTime != "" {
		where += " AND create_time <= ?"; args = append(args, query.EndTime)
	}

	orderCol := map[string]string{"username": "username", "operation": "operation", "executeTime": "execute_time", "createTime": "create_time"}[query.OrderBy]
	if orderCol == "" {
		orderCol = "create_time"
	}
	dir := "DESC"
	if query.Order == "asc" {
		dir = "ASC"
	}

	var total int64
	config.DB.QueryRow(fmt.Sprintf("SELECT COUNT(*) FROM operation_log %s", where), args...).Scan(&total)

	rows, err := config.DB.Query(
		fmt.Sprintf("SELECT id, username, operation, method, params, execute_time, ip, create_time FROM operation_log %s ORDER BY %s %s LIMIT ? OFFSET ?", where, orderCol, dir),
		append(args, query.PageSize, offset)...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	records := []models.OperationLog{}
	for rows.Next() {
		var l models.OperationLog
		rows.Scan(&l.ID, &l.Username, &l.Operation, &l.Method, &l.Params, &l.ExecuteTime, &l.IP, &l.CreateTime)
		l.Format()
		records = append(records, l)
	}

	return map[string]interface{}{"records": records, "total": total, "pageNum": query.PageNum, "pageSize": query.PageSize}, nil
}

// LogListAll 查询所有操作日志
func LogListAll() ([]models.OperationLog, error) {
	rows, err := config.DB.Query("SELECT id, username, operation, method, params, execute_time, ip, create_time FROM operation_log ORDER BY create_time DESC")
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	list := []models.OperationLog{}
	for rows.Next() {
		var l models.OperationLog
		rows.Scan(&l.ID, &l.Username, &l.Operation, &l.Method, &l.Params, &l.ExecuteTime, &l.IP, &l.CreateTime)
		l.Format()
		list = append(list, l)
	}
	return list, nil
}

// LogGetById 根据ID查询操作日志
func LogGetById(id string) (*models.OperationLog, error) {
	var l models.OperationLog
	err := config.DB.QueryRow("SELECT id, username, operation, method, params, execute_time, ip, create_time FROM operation_log WHERE id = ?", id).
		Scan(&l.ID, &l.Username, &l.Operation, &l.Method, &l.Params, &l.ExecuteTime, &l.IP, &l.CreateTime)
	if err != nil {
		return nil, &BizError{Code: utils.CodeNotFound, Message: "日志不存在"}
	}
	l.Format()
	return &l, nil
}

// LogDeleteById 删除操作日志
func LogDeleteById(id string) error {
	_, err := config.DB.Exec("DELETE FROM operation_log WHERE id = ?", id)
	return err
}

// LogDeleteBatch 批量删除操作日志
func LogDeleteBatch(ids []interface{}) error {
	if len(ids) == 0 {
		return &BizError{Code: utils.CodeBadRequest, Message: "参数错误"}
	}
	placeholders := strings.Repeat("?,", len(ids))
	placeholders = placeholders[:len(placeholders)-1]
	_, err := config.DB.Exec(fmt.Sprintf("DELETE FROM operation_log WHERE id IN (%s)", placeholders), ids...)
	return err
}
